File: Task/Palindrome-dates/C-sharp/palindrome-dates.cs
using System;
using System.Linq;
using System.Collections.Generic;

public class Program
{
    static void Main()
    {
        foreach (var date in PalindromicDates(2021).Take(15)) WriteLine(date.ToString("yyyy-MM-dd"));
    }

    public static IEnumerable<DateTime> PalindromicDates(int startYear) {
        for (int y = startYear; ; y++) {
            int m = Reverse(y % 100);
            int d = Reverse(y / 100);
            if (IsValidDate(y, m, d, out var date)) yield return date;
        }

        int Reverse(int x) => x % 10 * 10 + x / 10;
        bool IsValidDate(int y, int m, int d, out DateTime date) => DateTime.TryParse($"{y}-{m}-{d}", out date);
    }
}

File: Task/Palindrome-dates/PascalABC.NET/palindrome-dates.pas
uses system;

function Reverse(x: integer) := x mod 10 * 10 + x div 10;

function IsValidDate(y, m, d: integer; var date: datetime) :=
  DateTime.TryParse(y.tostring + '-' + m.tostring + '-' + d.tostring, date);

function PalindromicDates(startYear: integer): sequence of datetime;
begin
  var y := startyear;
  repeat
    var m := Reverse(y mod 100);
    var d := Reverse(y div 100);
    var date: datetime;
    if (IsValidDate(y, m, d, date)) then yield date;
    y += 1;
  until false;
end;

begin
  foreach var date in PalindromicDates(2021).Take(15) do
    Writeln(date.ToString('yyyy-MM-dd'));
end.
